clippy_lints/src/if_let_some_result.rs

```diff
@@ -39,19 +41,38 @@ declare_lint_pass!(OkIfLet => [IF_LET_SOME_RESULT]);
 impl<'a, 'tcx> LateLintPass<'a, 'tcx> for OkIfLet {
     fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'_>) {
         if_chain! { //begin checking variables
-            if let ExprKind::Match(ref op, ref body, ref source) = expr.kind; //test if expr is a match
-            if let MatchSource::IfLetDesugar { .. } = *source; //test if it is an If Let
-            if let ExprKind::MethodCall(_, _, ref result_types) = op.kind; //check is expr.ok() has type Result<T,E>.ok()
+            if let ExprKind::Match(ref op, ref body, source) = expr.kind; //test if expr is a match
+            if let MatchSource::IfLetDesugar { .. } = source; //test if it is an If Let
+            if let ExprKind::MethodCall(_, ok_span, ref result_types) = op.kind; //check is expr.ok() has type Result<T,E>.ok()
             if let PatKind::TupleStruct(QPath::Resolved(_, ref x), ref y, _)  = body[0].pat.kind; //get operation
             if method_chain_args(op, &["ok"]).is_some(); //test to see if using ok() methoduse std::marker::Sized;
 
             then {
                 let is_result_type = match_type(cx, cx.tables.expr_ty(&result_types[0]), &paths::RESULT);
-                let some_expr_string = snippet(cx, y[0].span, "");
+                let mut applicability = Applicability::MachineApplicable;
+                let trimed_ok_span = op.span.until(op.span.with_lo(ok_span.lo() - BytePos(1)));
+                let some_expr_string = snippet_with_applicability(cx, y[0].span, "", &mut applicability);
+                let trimmed_ok = snippet_with_applicability(cx, trimed_ok_span, "", &mut applicability);
+                let sugg = format!(
+                    "if let Ok({}) = {}",
+                    some_expr_string,
+                    trimmed_ok,
+                );
                 if print::to_string(print::NO_ANN, |s| s.print_path(x, false)) == "Some" && is_result_type {
-                    span_help_and_lint(cx, IF_LET_SOME_RESULT, expr.span,
-                    "Matching on `Some` with `ok()` is redundant",
-                    &format!("Consider matching on `Ok({})` and removing the call to `ok` instead", some_expr_string));
+                    span_lint_and_then(
+                        cx,
+                        IF_LET_SOME_RESULT,
+                        expr.span,
+                        "Matching on `Some` with `ok()` is redundant",
+                        |db| {
+                            db.span_suggestion(
+                                expr.span.shrink_to_lo().to(ok_span.with_hi(ok_span.hi() + BytePos(2))),
```

Comment:
This should then also be `expr.span.with_hi(ok_span.hi())`. 

This `BytePos` hack is hacky and won't work when a extra whitespace is added 
```rust
if let Some(_) = x.ok() {
// vs.
if let Some(_) = x.ok()  {
```

I think the `BytePos` thing I suggested above works, since even with whitespaces like this

```rust
if let Some(_) = x  . ok() {
// will be
if let Ok(_) = x   {
```

the dot `.` will be removed.

But I wouldn't trust me on this and add a test for this.

If it should break, you can either go back to the snippet string trimming, you had before, or get the parent expression and use `span.hi()` of the parent expression.

Reply:
Uhm, indeed I didn't think of such a strange pattern. `expr.span.with_hi(ok_span.hi())` seems to point to the correct place. I replaced it.
But it seems difficult to handle the second pattern. `op.span.with_lo(ok_span.lo() - BytePos(1))` returns ` ok()` (with one whitespace before it) so it looks like:

```rust
` ok()`
```
So I think it's an easy way to trim String. Added a commit.